proxy/upgrade_test.go

```diff
@@ -157,12 +170,18 @@ func TestServeHTTP(t *testing.T) {
 					return
 				}
 				defer conn.Close()
+
 				for {
 					s, err := bufrw.ReadString('\n')
-					if err != nil {
+					if err != nil && clientAlive() {
 						t.Errorf("error reading string: %v", err)
 						return
 					}
+
+					if ti.backendHangs {
```

Comment:
This if causes race check to fail although `ti` is used the same way several lines above without a problem :thinking:

Reply:
`ti := ti` seems to do the trick